File: ProgrammingProjects/cpp/common/src/vibrational_analysis.cpp
#include "chemistry/vibrational_analysis.h"
#include "chemistry/physical_constants.h"

#include <filesystem>
#include <fstream>
#include <utility>
#include <cmath>
#include <Eigen/Eigenvalues>
#include <stdexcept>
//
// Created by lrc on 2026/8/14.
//
namespace chemistry {

VibrationalAnalysis::VibrationalAnalysis(Molecule molecule, const std::filesystem::path& hessian_path):
_molecule(std::move(molecule)), _hessian(read_hessian(hessian_path, _molecule.atom_count()))
{
}
const Eigen::MatrixXd& VibrationalAnalysis::hessian() const {
    return _hessian;
}

Eigen::MatrixXd VibrationalAnalysis::read_hessian(const std::filesystem::path& path,
                                                  std::size_t expected_atom_count) {
    std::ifstream hessian_f(path);

    std::size_t file_atom_count;
    if (!(hessian_f >> file_atom_count)) {
        throw std::runtime_error("Can not open hessian file");
    }

    if (file_atom_count != expected_atom_count) {
        throw std::runtime_error("Atom count in Hessian file does not match expected number of atoms");
    }

    const Eigen::Index hessian_dim = 3 * static_cast<Eigen::Index>(expected_atom_count);

    Eigen::MatrixXd hessian(hessian_dim, hessian_dim);

    for (Eigen::Index i = 0; i < hessian_dim; ++i) {
        for (Eigen::Index j = 0; j < hessian_dim; ++j) {
            if (!(hessian_f >> hessian(i, j))) {
                throw std::runtime_error("Not enough values in Hessian file");
            }
        }
    }
    return hessian;
}

Eigen::MatrixXd VibrationalAnalysis::get_mass_weighted_hessian() const {
    int hessian_dim = 3 * _molecule.atom_count();
    Eigen::MatrixXd sqrt_mass_matrix(hessian_dim, hessian_dim);

    for (int i = 0; i < hessian_dim; ++i) {
        for (int j = 0; j < hessian_dim; ++j) {
            sqrt_mass_matrix(i, j) = sqrt(_molecule.atoms()[i/3].atomic_mass * _molecule.atoms()[j/3].atomic_mass) ;
        }
    }

    Eigen::MatrixXd mass_weighted_hessian = _hessian.cwiseQuotient(sqrt_mass_matrix);
    return mass_weighted_hessian;
}

NormalModeResult VibrationalAnalysis::get_normal_modes() const {
    const Eigen::MatrixXd mass_weighted_hessian = get_mass_weighted_hessian();

    Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> solver{mass_weighted_hessian};

    if (solver.info() != Eigen::Success) {
        throw std::runtime_error(
            "Failed to diagonalize mass-weighted Hessian"
        );
    }

    return NormalModeResult{solver.eigenvalues(), solver.eigenvectors()};
}

Eigen::VectorXd VibrationalAnalysis::get_wavenumbers_cm_inverse(
    double zero_tolerance) const {
    if (zero_tolerance < 0.0) {
        throw std::invalid_argument(
            "zero_tolerance must be non-negative"
        );
    }

    const Eigen::VectorXd eigenvalues =
        get_normal_modes().eigenvalues;

    const double conversion_factor =
        std::sqrt(
            hartree_to_joule
            / (amu_to_kilogram
               * bohr_to_meter
               * bohr_to_meter)
        )
        / (2.0 * pi * speed_of_light_cm_s);

    Eigen::VectorXd wavenumbers(eigenvalues.size());

    for (Eigen::Index index = 0;
         index < eigenvalues.size();
         ++index) {
        const double eigenvalue = eigenvalues(index);

        if (std::abs(eigenvalue) < zero_tolerance) {
            wavenumbers(index) = 0.0;
        } else {
            wavenumbers(index) = std::copysign(
                std::sqrt(std::abs(eigenvalue))
                    * conversion_factor,
                eigenvalue
            );
        }
    }

    return wavenumbers;
}

} // namespace chemistry

File: ProgrammingProjects/python/chemistry/vibrational_analysis.py
import numpy as np
from pathlib import Path

from .molecule import Molecule
from .physical_constants import (
    AMU_TO_KILOGRAM,
    BOHR_TO_METER,
    HARTREE_TO_JOULE,
    SPEED_OF_LIGHT_CM_S,
)


class VibrationalAnalysis:
    def __init__(self, molecule: Molecule, hessian_path):
        self._molecule = molecule
        self._hessian = VibrationalAnalysis.read_hessian(hessian_path, self._molecule.atom_count)

    @staticmethod
    def read_hessian(hessian_path: str | Path, expected_n_atom: int):
        hessian_f = open(hessian_path).readlines()
        file_n_atom = int(hessian_f[0])
        if file_n_atom != expected_n_atom:
            raise ValueError(f"Expected {expected_n_atom} atoms, got {file_n_atom}")

        values = []
        for line in hessian_f[1:]:
            values += [float(i) for i in line.strip().split()]
        values = np.array(values)

        hessian = np.reshape(values, (3 * expected_n_atom, 3 * expected_n_atom))
        return hessian

    @property
    def hessian(self):
        return self._hessian

    def get_mass_weighted_hessian(self):
        hessian = self._hessian
        molecule = self._molecule
        atoms = molecule.atoms
        sqrt_mass_matrix = np.zeros_like(hessian)
        for i in range(3 * molecule.atom_count):
            for j in range(3 * molecule.atom_count):
                sqrt_mass_matrix[i, j] = np.sqrt(atoms[i // 3].atomic_mass * atoms[j // 3].atomic_mass)

        mass_weighted_hessian = hessian / sqrt_mass_matrix
        return mass_weighted_hessian

    @property
    def mass_weighted_hessian(self):
        return self.get_mass_weighted_hessian()

    def get_normal_modes(self):
        mass_weighted_hessian = self.get_mass_weighted_hessian()

        eigenvalues, normal_modes = np.linalg.eigh(
            mass_weighted_hessian
        )

        return eigenvalues, normal_modes
    def get_wavenumbers_cm_inverse(
        self,
        zero_tolerance: float = 1e-10,
    ) -> np.ndarray:
        """Return signed normal-mode wavenumbers in cm^-1.

        The eigenvalues have units hartree / (amu * bohr^2).
        Converting that unit to SI gives angular-frequency squared:

            omega^2 = eigenvalue * E_h / (m_u * a_0^2)

        The spectroscopic wavenumber is omega / (2*pi*c), with c in
        cm/s. Significant negative eigenvalues are returned as negative
        wavenumbers, the usual notation for imaginary modes.
        """
        if zero_tolerance < 0.0:
            raise ValueError(
                "zero_tolerance must be non-negative"
            )

        eigenvalues, _ = self.get_normal_modes()

        conversion_factor = (
            np.sqrt(
                HARTREE_TO_JOULE
                / (
                    AMU_TO_KILOGRAM
                    * BOHR_TO_METER**2
                )
            )
            / (2.0 * np.pi * SPEED_OF_LIGHT_CM_S)
        )

        wavenumbers = (
            np.sign(eigenvalues)
            * np.sqrt(np.abs(eigenvalues))
            * conversion_factor
        )

        numerical_zero = np.abs(eigenvalues) < zero_tolerance
        wavenumbers[numerical_zero] = 0.0

        return wavenumbers
